abci/client/committing_client.go

```diff
@@ -22,17 +24,52 @@ type committingClient struct {
 	// Commit
 	// ApplySnapshotChunk
 	mtx *tmsync.RWMutex
+
+	// If called with false, we will wait until the app is ready to serve queries
+	// and deliveries.
+	// CONTRACT: The mtx read lock must be held.
+	//
+	// If called with true, we will indicate that the app is ready.
+	// CONTRACT: The mtx write lock must be held.
+	guardUntilStateIsCommitted func(bool)
```

Comment:
Consider bundling the `RWMutex`, the condition variable, and the underlying boolean into a single struct with `Lock/Unlock` and `RLock/RUnlock` methods which automatically implement the contract.

Reply:
Thanks, that's a good idea.